Fix autocomplete for

knudtty opened the pull request:
Added an override to provide fields as an input to SearchInputV2 (for lucene) and SqlInlineEditorControlled (for sql), instead of fetching in those respective components. This fixes autocomplete for fields on DBDashboardPage, which was not working previously. I have a fix for field values coming in a future PR.

![image](https://github.com/user-attachments/assets/93de4345-7e8d-4157-aaa1-d223ec49fc72)

![image](https://github.com/user-attachments/assets/c73eb00d-da0e-4fc3-adb4-3f9d2c600a70)

Ref: HDX-1471

knudtty commented on packages/app/src/DBDashboardPage.tsx:
```diff
@@ -506,6 +511,28 @@ function DBDashboardPage({ presetConfig }: { presetConfig?: Dashboard }) {
   const { data: sources } = useSources();
 
   const [highlightedTileId] = useQueryState('highlightedTileId');
+  const tableConnections = useMemo(() => {
```
Aggregates table connections for each tile and supplies the array into the lucene and sql where inputs

knudtty commented on packages/app/src/DBDashboardPage.tsx:
```diff
@@ -900,9 +918,7 @@ function DBDashboardPage({ presetConfig }: { presetConfig?: Dashboard }) {
-                connectionId={defaultSource?.connection}
-                database={defaultDatabaseName}
-                table={defaultTableName}
+                tableConnection={tableConnections}
```
You'll see a lot of this, I combined databaseName, tableName, and connectionId into an object as they are commonly needed for many metadata operations.

knudtty commented on packages/app/src/DBSearchPage.tsx:
```diff
@@ -20,6 +20,7 @@ import { useForm } from 'react-hook-form';
 import { z } from 'zod';
 import { zodResolver } from '@hookform/resolvers/zod';
 import { ClickHouseQueryError } from '@hyperdx/common-utils/dist/clickhouse';
+import { tcFromSource } from '@hyperdx/common-utils/dist/metadata';
```
Helper function to build a TableConnection from a TSource

knudtty commented on packages/app/src/SearchInputV2.tsx:
```diff
@@ -38,16 +38,15 @@ export default function SearchInputV2({
+      : isTableConnection(_tableConnections)
+        ? [_tableConnections]
+        : _tableConnections;
+  const { data: fields } = useAllFields(tableConnections, {
```
This hook now accepts an either a single TableConnection or an array of them

wrn14897 replied:
can you explain more why do we want to accept either a single TableConnection or an array of them? Can't we just do array?

knudtty commented on packages/app/src/hooks/useMetadata.tsx:
```diff
@@ -36,26 +41,20 @@ export function useColumns(
-    tableName: string;
-    connectionId: string;
-  },
+  _tableConnections: TableConnection | TableConnection[],
```
refactors to support one-to-many TableConnections

knudtty commented on packages/common-utils/src/metadata.ts:
```diff
@@ -349,53 +349,65 @@ export class Metadata {
-    tableName: string;
-    connectionId: string;
-  }) {
+  async getAllFields(...tableConnections: TableConnection[]) {
```
Significant refactor to allow fetching fields for many tables

knudtty commented on packages/common-utils/src/metadata.ts:
```diff
@@ -467,6 +479,37 @@ export type Field = {
   jsType: JSDataType | null;
 };
 
+export type TableConnection = {
```
New type w/ helper functions below

wrn14897 reviewed (changes requested):
Its great to have new `TableConnection` type. I'm a bit worried about the refactoring of the metadata file where `getAllFields` plays a critical role across the app. I'd suggest to move refactoring parts to another PR (ideally after next release). For this PR, we can just hot fix the autocomplete part

knudtty commented:
Yeah I didn't like how much modification I did to getAllFields as well. I refactored so there are now no changes to the function, just a little deduplication logic I added to useAllFields. Functionally it does the same thing, but a far less intrusive change.

wrn14897 commented on packages/app/src/hooks/useMetadata.tsx:
```diff
@@ -36,26 +41,39 @@ export function useColumns(
-        tableName,
-        connectionId,
-      });
+      const metadata = getMetadata();
```
we better leave the metadata (caching singleton) init outside the function so its not reinit for every queryFn calls

knudtty replied:
Cool, I moved it inside because I saw it being done in some other hooks. We should probably do this across the board at some point.

wrn14897 replied:
Good point! I noticed that and I will do the clean up in a separate PR.

wrn14897 commented on packages/app/src/hooks/useMetadata.tsx:
```diff
@@ -36,26 +41,39 @@ export function useColumns(
+      const set = new Set<string>();
+      for (const _fields of fields2d) {
+        for (const field of _fields) {
+          const key = `${field.path.join('.')}_${field.jsType?.toString()}`;
```
can just use `objectHash` 
```
const _key = objectHash.sha1(field)
```

wrn14897 commented on packages/app/src/hooks/useMetadata.tsx:
```diff
@@ -36,26 +41,39 @@ export function useColumns(
+      // skip deduplication if not possible
+      if (fields2d.length === 1) return fields2d[0];
+
+      // deduplicate common fields
+      const fields = [];
+      const set = new Set<string>();
+      for (const _fields of fields2d) {
+        for (const field of _fields) {
+          const key = `${field.path.join('.')}_${field.jsType?.toString()}`;
```
is it possible to add some unit tests for `useAllFields`  ?

knudtty replied:
I'll extract the deduplication into a function and add unit tests for that. For a full test of the useAllFields hook, I think we need to come up with a strategy to spin up a test CH instance for testing w/ seed data.

wrn14897 commented:
> Yeah I didn't like how much modification I did to getAllFields as well. I refactored so there are now no changes to the function, just a little deduplication logic I added to useAllFields. Functionally it does the same thing, but a far less intrusive change.

Sounds good. Before the next release, I'd prefer us to touch less on `common-utils` if possible :)

wrn14897 commented on packages/app/src/DBDashboardPage.tsx:
```diff
@@ -506,6 +511,28 @@ function DBDashboardPage({ presetConfig }: { presetConfig?: Dashboard }) {
+    for (const { config } of dashboard.tiles) {
+      const source = sources?.find(v => v.id === config.source);
+      if (!source) continue;
+      // TODO: will need to update this when we allow for multiple metrics per chart
```
theoretically, we can go through each select to pull the table name, right? fwiw, there will be only 3 tables total for a single source (you can find them in `source.metricTables`)

wrn14897 replied:
I think its okay to pull the 1st one for now

knudtty replied:
Probably - this is where I'm more of a novice still. Some of this is table fetching logic is from another hook in this file, I copied over the comment from there so it's not left behind. I'll leave it as is for now, if there's a change you're thinking lmk!

knudtty commented on packages/app/src/DBDashboardPage.tsx:
```diff
@@ -886,9 +906,7 @@ function DBDashboardPage({ presetConfig }: { presetConfig?: Dashboard }) {
-                connectionId={defaultSource?.connection}
-                database={defaultDatabaseName}
-                table={defaultTableName}
+                tableConnection={tableConnections}
```
Makes sense, I renamed it to tableConnections. I feel that's more sensible, even if you're only passing one connection.

wrn14897 commented on packages/app/src/hooks/__tests__/useMetadata.test.tsx:
```diff
@@ -0,0 +1,107 @@
+  });
+
+  // Test performance consideration (large arrays)
+  it('should handle large 2D arrays efficiently', () => {
```
Lol...like test until it timeout or what?

knudtty replied:
I thought I removed that one tbh. Out now!

wrn14897 reviewed (approved):
lgtm

wrn14897 commented on packages/app/src/hooks/useMetadata.tsx:
```diff
@@ -115,3 +122,21 @@ export function useGetKeyValues({
+  return array;
+}
+
+// export functions for testing only
+export const testExports =
+  process.env.NODE_ENV === 'test' ? { deduplicate2dArray } : undefined;
```
we probably don't set the node env for unit test (also you want to use `IS_CI` flag instead). I think we can just export this function
